### PR TITLE
feat: v5- polyline

### PR DESCRIPTION
**Pull request by yvonneyx**
Feature:  Polyline

目前实现以下功能点：

1. 继承 line 的基本功能

Kapture 2023-08-09 at 20 39 42

2. 参数化绘制：使用控制点和转角度数进行绘制

image

3. A* 寻径算法集成：在没有指定控制点的情况下，通过 Waypoints 方式求得最短路径

image 
image

4. 自动避障功能：边的路由配置 `routeCfg` 中支持配置 `obstacleAvoidance` 属性

Kapture 2023-08-09 at 21 02 57

5. 防止边重叠：点支持配置 `preventEdgeOverlap` 属性，确保节点在移动的过程中，点可以抢占其他边的位置，避免重叠

Kapture 2023-08-09 at 21 04 55

性能测试：PFS

<ne-clipboard data="%7B%22type%22%3A%22fragment%22%2C%22name%22%3A%22%23fragment%22%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22ue354176a%22%2C%22name%22%3A%22tableHole%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22jc1Kq%22%2C%22name%22%3A%22table%22%2C%22attrs%22%3A%7B%22spacing%22%3Atrue%2C%22widthMode%22%3A%22contain%22%2C%22colWidths%22%3A%5B231%2C231%2C233%5D%2C%22rowCount%22%3A3%2C%22colCount%22%3A3%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u452d094f%22%2C%22name%22%3A%22tr%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u507698d0%22%2C%22name%22%3A%22td%22%2C%22attrs%22%3A%7B%22col%22%3A0%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u87e4da1a%22%2C%22name%22%3A%22p%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22text%22%2C%22id%22%3A%22ub1a38709%22%2C%22name%22%3A%22%23text%22%2C%22attrs%22%3A%7B%7D%2C%22data%22%3A%22%22%7D%5D%7D%5D%7D%2C%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u62583b0f%22%2C%22name%22%3A%22td%22%2C%22attrs%22%3A%7B%22col%22%3A1%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u593c4f1f%22%2C%22name%22%3A%22p%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22text%22%2C%22id%22%3A%22u51e4b101%22%2C%22name%22%3A%22%23text%22%2C%22attrs%22%3A%7B%7D%2C%22data%22%3A%22100%E4%B8%AA%E7%82%B9100%E6%9D%A1%E8%BE%B9%22%7D%5D%7D%5D%7D%2C%7B%22type%22%3A%22element%22%2C%22id%22%3A%22udd00a7f7%22%2C%22name%22%3A%22td%22%2C%22attrs%22%3A%7B%22col%22%3A2%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22ua57d0409%22%2C%22name%22%3A%22p%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22text%22%2C%22id%22%3A%22ubce6f359%22%2C%22name%22%3A%22%23text%22%2C%22attrs%22%3A%7B%7D%2C%22data%22%3A%221000%E4%B8%AA%E7%82%B91000%E6%9D%A1%E8%BE%B9%22%7D%5D%7D%5D%7D%5D%7D%2C%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u56718ec3%22%2C%22name%22%3A%22tr%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22ufccfec53%22%2C%22name%22%3A%22td%22%2C%22attrs%22%3A%7B%22col%22%3A0%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u1e8afac6%22%2C%22name%22%3A%22p%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22text%22%2C%22id%22%3A%22uf81a4adb%22%2C%22name%22%3A%22%23text%22%2C%22attrs%22%3A%7B%7D%2C%22data%22%3A%22%E9%A6%96%E5%B1%8F%E6%B8%B2%E6%9F%93%E6%97%B6%E9%97%B4%22%7D%5D%7D%5D%7D%2C%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u6813e406%22%2C%22name%22%3A%22td%22%2C%22attrs%22%3A%7B%22col%22%3A1%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u48c9e232%22%2C%22name%22%3A%22p%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22text%22%2C%22id%22%3A%22u998179aa%22%2C%22name%22%3A%22%23text%22%2C%22attrs%22%3A%7B%7D%2C%22data%22%3A%22212ms%22%7D%5D%7D%5D%7D%2C%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u58e8f164%22%2C%22name%22%3A%22td%22%2C%22attrs%22%3A%7B%22col%22%3A2%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22ud98f53ff%22%2C%22name%22%3A%22p%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22text%22%2C%22id%22%3A%22ub64faeb6%22%2C%22name%22%3A%22%23text%22%2C%22attrs%22%3A%7B%7D%2C%22data%22%3A%22223ms%22%7D%5D%7D%5D%7D%5D%7D%2C%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u46e60efa%22%2C%22name%22%3A%22tr%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22ua83ef7e7%22%2C%22name%22%3A%22td%22%2C%22attrs%22%3A%7B%22col%22%3A0%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u5a692a24%22%2C%22name%22%3A%22p%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22text%22%2C%22id%22%3A%22u33a702b4%22%2C%22name%22%3A%22%23text%22%2C%22attrs%22%3A%7B%7D%2C%22data%22%3A%22%E9%A6%96%E5%B1%8F%E6%B8%B2%E6%9F%93%E6%97%B6%E9%97%B4%EF%BC%88%E5%BC%80%E5%90%AF%E8%87%AA%E5%8A%A8%E9%81%BF%E9%9A%9C%EF%BC%89%22%7D%5D%7D%5D%7D%2C%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u961a8ca4%22%2C%22name%22%3A%22td%22%2C%22attrs%22%3A%7B%22col%22%3A1%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22uad3837e3%22%2C%22name%22%3A%22p%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22text%22%2C%22id%22%3A%22uebcfa1a5%22%2C%22name%22%3A%22%23text%22%2C%22attrs%22%3A%7B%7D%2C%22data%22%3A%22208ms%22%7D%5D%7D%5D%7D%2C%7B%22type%22%3A%22element%22%2C%22id%22%3A%22ub188d850%22%2C%22name%22%3A%22td%22%2C%22attrs%22%3A%7B%22col%22%3A2%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22element%22%2C%22id%22%3A%22u62d02382%22%2C%22name%22%3A%22p%22%2C%22attrs%22%3A%7B%7D%2C%22children%22%3A%5B%7B%22type%22%3A%22text%22%2C%22id%22%3A%22u20e8c392%22%2C%22name%22%3A%22%23text%22%2C%22attrs%22%3A%7B%7D%2C%22data%22%3A%22471ms%22%7D%5D%7D%5D%7D%5D%7D%5D%7D%5D%7D%5D%2C%22attrs%22%3A%7B%7D%7D"></ne-clipboard><div class="lake-content" typography="classic">

  | 100个点100条边 | 1000个点1000条边
-- | -- | --
首屏渲染时间 | 212ms | 223ms
首屏渲染时间（开启自动避障） | 208ms | 471ms

</div>

TODO:

tests

### CODE REVIEW

File: packages/g6/src/runtime/graph.ts [hunk 1]
```diff
@@ -851,6 +852,18 @@ export default class Graph<B extends BehaviorRegistry, T extends ThemeRegistry>
     return this.dataController.findChildren(comboId, 'combo');
   }
 
+  /**
+   * Identify edges that are affected when a particular node is moved
+   * @param nodeId id of node that is moving
+   * @returns list of affected edges(including related edges and closest edges)
+   */
+  public getAffectedEdgesByNodeMovement(nodeId: ID): EdgeModel[] {
```
Comment (Yanyan-Wang) [inline packages/g6/src/runtime/graph.ts:860]: 这个为什么需要放在 graph 的 API 上呀，理论上 graph 上的 API 是希望外部用户用的哈。是不是因为需要 data 和 item controller 的相互协作？看起来是不是 dataController 里面这个方法 findAffectedEdgesByNodeMovement 可以在 item 里面实现，它也比较像 item 的一下操作。这样的话就 item 调用自己的函数就好了，不用通过 graph

File: packages/g6/src/types/node.ts [hunk 2]
```diff
@@ -68,6 +68,12 @@ export interface NodeUserModelData extends PlainObject {
     text: string;
     position: BadgePosition;
   }[];
+  /**
+   * Whether to prevent overlap with unassociated edges. Used to preempt position.
+   * Defaults to false.
+   * Only valid for polyline
+   */
+  preventEdgeOverlap?: boolean;
```
Comment (Yanyan-Wang) [inline packages/g6/src/types/node.ts:76]: 这个参数只对 polyline 生效的话，要不更明确一些的名字，preventPolylineEdgeOverlap

Comment (yvonneyx) [reply packages/g6/src/types/node.ts:76]: 已修改

File: packages/g6/src/runtime/controller/item.ts [hunk 3]
```diff
@@ -1203,3 +1219,18 @@ const getItemTheme = (
     lodStrategy: formattedLodStrategy,
   };
 };
+
+export const filterVisibleItemMapByType = (
```
[the remark this answers is not in the document]
Comment (yvonneyx) [reply packages/g6/src/runtime/controller/item.ts:1223]: 仅在改文件中使用，删除 `export`

File: packages/g6/src/runtime/graph.ts (hunk 1 above)
Comment (yvonneyx) [reply packages/g6/src/runtime/graph.ts:860]: 在 drag-node 交互中需要用到这个函数，现在改在 item controller 里面实现了该方法。如果不借助 graph API 将方法暴露出来，怎么获取呢？

File: packages/g6/src/types/node.ts (hunk 2 above)
Comment (Yanyan-Wang) [reply packages/g6/src/types/node.ts:76]: 是不是修改还没有 push ～

File: packages/g6/src/runtime/graph.ts (hunk 1 above)
Comment (Yanyan-Wang) [reply packages/g6/src/runtime/graph.ts:860]: 嗷那如果 drag-node 要用好像确实需要暴露。这个函数的意思是检测和 nodeId 有 overlap 的边么？

Comment (yvonneyx) [reply packages/g6/src/runtime/graph.ts]: `graph` adds `getNearEdgesForNode` API to retrieve the nearby edges for a given node using quadtree collision detection.

File: packages/g6/src/stdlib/behavior/drag-node.ts [hunk 4]
```diff
@@ -294,6 +313,31 @@ export default class DragNode extends Behavior {
       this.originY = event.canvas.y;
     }
 
+    if (this.dragging && enableTransient) {
```
[the remark this answers is not in the document]
Comment (yvonneyx) [reply packages/g6/src/stdlib/behavior/drag-node.ts:323]: When dragging nodes, if nodes are set to `preventPolylineEdgeOverlap`, use quadtree collision detection to identity nearby edges and dynamically update them

File: packages/g6/src/runtime/controller/item.ts [hunk 5]
```diff
@@ -995,6 +1005,11 @@ export class ItemController {
         edgeTheme,
       );
 
+      const nodeMap = filterVisibleItemMapByType(
```
[the remark this answers is not in the document]
Comment (yvonneyx) [reply packages/g6/src/runtime/controller/item.ts]: 这块确实没考虑到，现在调整为在动态规划路线时过滤出 visible 为 true 的点作为障碍

**Review by Yanyan-Wang — APPROVED** [review]

**Review by pomelo-nwu — APPROVED** [review]
LGTM

**Pull request merged**